File: assets/csharp/run-with-it-pr-body.cs
#!/usr/bin/env dotnet
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

class Program
{
    static readonly Regex AutoCloseRefPattern = new(
        @"\b(close(?:s|d)?|fix(?:es|ed)?|resolve(?:s|d)?)(\s+)((?:[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)?)(#\d+)\b",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant
    );

    static int Main(string[] args)
    {
        if (args.Length == 0)
        {
            PrintUsage();
            return 2;
        }

        if (args.Contains("--help") || args.Contains("-h"))
        {
            PrintUsage();
            return 0;
        }

        var command = args[0];
        if (command != "render")
        {
            PrintUsage();
            return 2;
        }

        string stateFile = "";
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i] == "--state-file" && i + 1 < args.Length)
            {
                stateFile = args[i + 1];
                i++;
            }
            else
            {
                PrintUsage();
                return 2;
            }
        }

        if (string.IsNullOrEmpty(stateFile))
        {
            PrintUsage();
            return 2;
        }

        try
        {
            Console.Write(RenderPrBody(stateFile));
            return 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex.Message);
            return 2;
        }
    }

    static string RenderPrBody(string stateFile)
    {
        var state = LoadState(stateFile);
        var runRoot = RunRootFor(stateFile);
        var counts = StatusCounts(state);
        var summaries = SummaryByIssue(state);
        var completed = CompletedIssueNumbers(state);

        int totalAdded = 0;
        int totalDeleted = 0;
        foreach (var issue in completed)
        {
            if (summaries.TryGetValue(issue, out var sum))
            {
                totalAdded += AsInt(sum["lines_added"]) ?? 0;
                totalDeleted += AsInt(sum["lines_deleted"]) ?? 0;
            }
        }

        var lines = new List<string>
        {
            "## Summary",
            $"- Total issues processed: {counts.Values.Sum()}",
            $"- Completed: {counts["completed"]}",
            $"- Failed review: {counts["failed-review"]}",
            $"- Failed merge: {counts["failed-merge"]}",
            $"- Blocked: {counts["blocked"]}",
            $"- Lines added: {totalAdded}",
            $"- Lines deleted: {totalDeleted}",
            "",
            "## Closed Issues"
        };

        if (completed.Count > 0)
        {
            foreach (var issue in completed)
            {
                lines.Add($"- #{issue}");
            }
        }
        else
        {
            lines.Add("None");
        }

        lines.Add("");
        lines.Add("## Models Used");
        lines.Add("| Issue | Task | Cycle | Agent | Model | Reason |");
        lines.Add("|---|---|---:|---|---|---|");

        foreach (var issue in completed)
        {
            var summary = summaries.TryGetValue(issue, out var sum) ? sum : new JsonObject();
            var report = ReportForIssue(state, issue, runRoot, summary);
            foreach (var row in ModelRows(issue, report, summary))
            {
                lines.Add(
                    $"| #{OneLine(issue)} | {row["role"]} | {row["cycle"]} | " +
                    $"{row["agent"]} | {row["model"]} | {row["selection_reason"]} |"
                );
            }
        }

        lines.Add("");
        lines.Add("## Verification");
        lines.Add("| Issue | State | Evidence |");
        lines.Add("|---|---|---|");

        foreach (var issue in completed)
        {
            var summary = summaries.TryGetValue(issue, out var sum) ? sum : new JsonObject();
            var report = ReportForIssue(state, issue, runRoot, summary);
            var (stateText, evidence) = VerificationState(report, summary);
            lines.Add($"| #{OneLine(issue)} | {stateText} | {evidence} |");
        }

        return string.Join("\n", lines) + "\n";
    }

    static JsonObject LoadState(string path)
    {
        try
        {
            if (!File.Exists(path))
            {
                Console.Error.WriteLine($"error: failed to load state file {path}: File not found");
                Environment.Exit(1);
            }
            var text = File.ReadAllText(path);
            var node = JsonNode.Parse(text);
            if (node is JsonObject obj)
            {
                return obj;
            }
            Console.Error.WriteLine($"error: state file {path} must contain a JSON object");
            Environment.Exit(1);
            return null!;
        }
        catch (Exception exc)
        {
            Console.Error.WriteLine($"error: failed to load state file {path}: {exc.Message}");
            Environment.Exit(1);
            return null!;
        }
    }

    static string RunRootFor(string stateFile)
    {
        var fullPath = Path.GetFullPath(stateFile);
        var parent = Path.GetDirectoryName(fullPath);
        if (parent != null && Path.GetFileName(parent) == ".run-with-it")
        {
            return Path.GetDirectoryName(parent) ?? parent;
        }
        return parent ?? "";
    }

    static string ResolvePath(string? path, string runRoot)
    {
        if (string.IsNullOrEmpty(path)) return "";
        if (Path.IsPathRooted(path)) return path;
        return Path.GetFullPath(Path.Combine(runRoot, path));
    }

    static string OneLine(string? value)
    {
        var text = string.IsNullOrEmpty(value) ? "unknown" : value;
        text = text.Replace("|", "\\|");
        var parts = text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        text = string.Join(" ", parts);
        return SanitizeAutoCloseRefs(text);
    }

    static string SanitizeAutoCloseRefs(string text)
    {
        return AutoCloseRefPattern.Replace(text, match => {
            return $"{match.Groups[1].Value}{match.Groups[2].Value}{match.Groups[3].Value}\\{match.Groups[4].Value}";
        });
    }

    static Dictionary<string, int> StatusCounts(JsonObject state)
    {
        var counts = new Dictionary<string, int>(StringComparer.Ordinal)
        {
            ["completed"] = 0,
            ["failed-review"] = 0,
            ["failed-merge"] = 0,
            ["blocked"] = 0
        };
        if (state.TryGetPropertyValue("issue_registry", out var regNode) && regNode is JsonObject registry)
        {
            foreach (var pair in registry)
            {
                if (pair.Value is JsonObject info)
                {
                    var status = AsString(info["status"]) ?? "";
                    if (counts.ContainsKey(status))
                    {
                        counts[status]++;
                    }
                }
            }
        }
        return counts;
    }

    static IEnumerable<string> SortIssues(IEnumerable<string> issues)
    {
        return issues.OrderBy(issue => {
            if (int.TryParse(issue, NumberStyles.Integer, CultureInfo.InvariantCulture, out var val))
            {
                return (0, val.ToString("D20", CultureInfo.InvariantCulture));
            }
            return (1, issue);
        }, Comparer<(int, string)>.Default);
    }

    static List<string> CompletedIssueNumbers(JsonObject state)
    {
        var list = new List<string>();
        if (state.TryGetPropertyValue("issue_registry", out var regNode) && regNode is JsonObject registry)
        {
            foreach (var pair in registry)
            {
                if (pair.Value is JsonObject info)
                {
                    if (AsString(info["status"]) == "completed")
                    {
                        list.Add(pair.Key);
                    }
                }
            }
        }
        return SortIssues(list).ToList();
    }

    static Dictionary<string, JsonObject> SummaryByIssue(JsonObject state)
    {
        var result = new Dictionary<string, JsonObject>(StringComparer.Ordinal);
        if (state.TryGetPropertyValue("completed_summaries", out var sumNode) && sumNode is JsonArray summaries)
        {
            foreach (var item in summaries)
            {
                if (item is JsonObject obj && obj.TryGetPropertyValue("issue", out var issueVal) && issueVal != null)
                {
                    var issueStr = AsIntText(issueVal) ?? issueVal.ToString();
                    result[issueStr] = obj;
                }
            }
        }
        return result;
    }

    static JsonObject ReportForIssue(JsonObject state, string issue, string runRoot, JsonObject summary)
    {
        var summaryReport = ResolvePath(AsString(summary["report_file"]), runRoot);
        if (!string.IsNullOrEmpty(summaryReport) && File.Exists(summaryReport))
        {
            return LoadJson(summaryReport);
        }

        if (state.TryGetPropertyValue("issue_registry", out var regNode) && regNode is JsonObject registry)
        {
            if (registry.TryGetPropertyValue(issue, out var infoNode) && infoNode is JsonObject info)
            {
                var refs = new[] { AsString(info["merge_recovery_report_file"]), AsString(info["report_file"]) };
                foreach (var r in refs)
                {
                    var reportPath = ResolvePath(r, runRoot);
                    if (!string.IsNullOrEmpty(reportPath) && File.Exists(reportPath))
                    {
                        var report = LoadJson(reportPath);
                        if (report.Count > 0) return report;
                    }
                }
            }
        }
        return new JsonObject();
    }

    static List<JsonObject> ModelRows(string issue, JsonObject report, JsonObject summary)
    {
        var usage = report["model_usage"] as JsonArray ?? summary["model_usage"] as JsonArray;
        if (usage == null || usage.Count == 0)
        {
            return new List<JsonObject>
            {
                new JsonObject
                {
                    ["issue"] = issue,
                    ["role"] = "unknown",
                    ["cycle"] = "-",
                    ["agent"] = "unknown",
                    ["model"] = "unknown",
                    ["selection_reason"] = "missing-model-usage",
                }
            };
        }

        var rows = new List<JsonObject>();
        foreach (var item in usage)
        {
            if (item is JsonObject row)
            {
                var cycleVal = row.TryGetPropertyValue("cycle", out var cVal) && cVal != null ? AsIntText(cVal) ?? cVal.ToString() : "-";
                rows.Add(new JsonObject
                {
                    ["issue"] = issue,
                    ["role"] = OneLine(AsString(row["role"])),
                    ["cycle"] = OneLine(cycleVal),
                    ["agent"] = OneLine(AsString(row["agent"])),
                    ["model"] = OneLine(AsString(row["model"])),
                    ["selection_reason"] = OneLine(AsString(row["selection_reason"]) ?? AsString(row["reason"])),
                });
            }
        }

        if (rows.Count == 0)
        {
            return ModelRows(issue, new JsonObject(), new JsonObject());
        }
        return rows;
    }

    static (string State, string Evidence) VerificationState(JsonObject report, JsonObject summary)
    {
        var verification = report["verification"] as JsonObject ?? summary["verification"] as JsonObject ?? new JsonObject();

        string state = "unknown";
        if (verification.TryGetPropertyValue("passed", out var passedNode) && passedNode is JsonValue passedVal)
        {
            if (passedVal.GetValueKind() == JsonValueKind.True) state = "passed";
            else if (passedVal.GetValueKind() == JsonValueKind.False) state = "failed";
        }

        var evidence = AsString(verification["evidence"]) ?? AsString(summary["summary"]) ?? "unknown";
        return (state, OneLine(evidence));
    }

    static JsonObject LoadJson(string path)
    {
        if (!File.Exists(path) || new FileInfo(path).Length == 0) return new JsonObject();
        try
        {
            var text = File.ReadAllText(path);
            return JsonNode.Parse(text) as JsonObject ?? new JsonObject();
        }
        catch
        {
            return new JsonObject();
        }
    }

    static string? AsString(JsonNode? node)
    {
        return node is JsonValue value && value.GetValueKind() == JsonValueKind.String
            ? value.GetValue<string>()
            : null;
    }

    static int? AsInt(JsonNode? node)
    {
        if (node is JsonValue value && value.GetValueKind() == JsonValueKind.Number)
        {
            if (value.TryGetValue<int>(out var parsed))
            {
                return parsed;
            }
            if (value.TryGetValue<double>(out var dbl))
            {
                return (int)dbl;
            }
        }
        return null;
    }

    static string? AsIntText(JsonNode? node)
    {
        if (node is JsonValue value)
        {
            if (value.GetValueKind() == JsonValueKind.Number)
            {
                if (value.TryGetValue<int>(out var intValue))
                {
                    return intValue.ToString(CultureInfo.InvariantCulture);
                }
                if (value.TryGetValue<double>(out var doubleValue))
                {
                    return doubleValue.ToString(CultureInfo.InvariantCulture);
                }
            }
            if (value.GetValueKind() == JsonValueKind.String)
            {
                return value.GetValue<string>();
            }
        }
        return null;
    }

    static void PrintUsage()
    {
        Console.WriteLine("Usage: run-with-it-pr-body render --state-file <file>");
    }
}

File: assets/python/run-with-it-pr-body.py
#!/usr/bin/env python3
"""Render final run-with-it pull request bodies from compact state."""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any

AUTO_CLOSE_REF_PATTERN = re.compile(
    r"\b(close(?:s|d)?|fix(?:es|ed)?|resolve(?:s|d)?)(\s+)((?:[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)?)(#\d+)\b",
    re.IGNORECASE,
)


def load_json(path: str) -> dict[str, Any]:
    try:
        with open(path, "r", encoding="utf-8") as handle:
            value = json.load(handle)
        return value if isinstance(value, dict) else {}
    except Exception:
        return {}


def load_state(path: str) -> dict[str, Any]:
    try:
        with open(path, "r", encoding="utf-8") as handle:
            value = json.load(handle)
    except Exception as exc:
        raise SystemExit(f"error: failed to load state file {path}: {exc}") from exc
    if not isinstance(value, dict):
        raise SystemExit(f"error: state file {path} must contain a JSON object")
    return value


def run_root_for(state_file: str) -> Path:
    state_path = Path(state_file).resolve()
    if state_path.parent.name == ".run-with-it":
        return state_path.parent.parent
    return state_path.parent


def resolve_path(path: Any, run_root: Path) -> str:
    if not isinstance(path, str) or not path:
        return ""
    candidate = Path(path)
    if candidate.is_absolute():
        return str(candidate)
    return str(run_root / candidate)


def issue_sort_key(value: Any) -> tuple[int, str]:
    text = str(value)
    try:
        return (0, f"{int(text):020d}")
    except ValueError:
        return (1, text)


def one_line(value: Any) -> str:
    text = "unknown" if value is None or value == "" else str(value)
    text = " ".join(text.replace("|", "\\|").split())
    return sanitize_auto_close_refs(text)


def sanitize_auto_close_refs(text: str) -> str:
    return AUTO_CLOSE_REF_PATTERN.sub(lambda match: f"{match[1]}{match[2]}{match[3]}\\{match[4]}", text)


def status_counts(state: dict[str, Any]) -> dict[str, int]:
    counts = {"completed": 0, "failed-review": 0, "failed-merge": 0, "blocked": 0}
    for info in state.get("issue_registry", {}).values():
        if isinstance(info, dict):
            status = str(info.get("status", ""))
            if status in counts:
                counts[status] += 1
    return counts


def completed_issue_numbers(state: dict[str, Any]) -> list[str]:
    registry = state.get("issue_registry", {})
    issues = [
        str(issue)
        for issue, info in registry.items()
        if isinstance(info, dict) and info.get("status") == "completed"
    ]
    return sorted(issues, key=issue_sort_key)


def summary_by_issue(state: dict[str, Any]) -> dict[str, dict[str, Any]]:
    result: dict[str, dict[str, Any]] = {}
    for item in state.get("completed_summaries", []):
        if isinstance(item, dict) and item.get("issue") is not None:
            result[str(item["issue"])] = item
    return result


def report_for_issue(
    state: dict[str, Any], issue: str, run_root: Path, summary: dict[str, Any]
) -> dict[str, Any]:
    summary_report = resolve_path(summary.get("report_file"), run_root)
    if summary_report:
        return load_json(summary_report)

    info = state.get("issue_registry", {}).get(str(issue), {})
    if not isinstance(info, dict):
        info = {}
    for report_ref in (info.get("merge_recovery_report_file"), info.get("report_file")):
        report_path = resolve_path(report_ref, run_root)
        if report_path:
            report = load_json(report_path)
            if report:
                return report
    return {}


def model_rows(issue: str, report: dict[str, Any], summary: dict[str, Any]) -> list[dict[str, Any]]:
    usage = report.get("model_usage")
    if not isinstance(usage, list):
        usage = summary.get("model_usage")
    if not isinstance(usage, list) or not usage:
        return [
            {
                "issue": issue,
                "role": "unknown",
                "cycle": "-",
                "agent": "unknown",
                "model": "unknown",
                "selection_reason": "missing-model-usage",
            }
        ]

    rows: list[dict[str, Any]] = []
    for item in usage:
        if not isinstance(item, dict):
            continue
        rows.append(
            {
                "issue": issue,
                "role": one_line(item.get("role")),
                "cycle": one_line(item.get("cycle") if item.get("cycle") is not None else "-"),
                "agent": one_line(item.get("agent")),
                "model": one_line(item.get("model")),
                "selection_reason": one_line(item.get("selection_reason") or item.get("reason")),
            }
        )
    return rows or model_rows(issue, {}, {})


def verification_state(report: dict[str, Any], summary: dict[str, Any]) -> tuple[str, str]:
    verification = report.get("verification")
    if not isinstance(verification, dict):
        summary_verification = summary.get("verification")
        verification = summary_verification if isinstance(summary_verification, dict) else {}
    passed = verification.get("passed")
    state = "passed" if passed is True else "failed" if passed is False else "unknown"
    evidence = verification.get("evidence") or summary.get("summary") or "unknown"
    return state, one_line(evidence)


def render_pr_body(state_file: str) -> str:
    state = load_state(state_file)
    run_root = run_root_for(state_file)
    counts = status_counts(state)
    summaries = summary_by_issue(state)
    completed = completed_issue_numbers(state)
    total_added = sum(int(summaries.get(issue, {}).get("lines_added", 0) or 0) for issue in completed)
    total_deleted = sum(int(summaries.get(issue, {}).get("lines_deleted", 0) or 0) for issue in completed)

    lines = [
        "## Summary",
        f"- Total issues processed: {sum(counts.values())}",
        f"- Completed: {counts['completed']}",
        f"- Failed review: {counts['failed-review']}",
        f"- Failed merge: {counts['failed-merge']}",
        f"- Blocked: {counts['blocked']}",
        f"- Lines added: {total_added}",
        f"- Lines deleted: {total_deleted}",
        "",
        "## Closed Issues",
    ]

    if completed:
        lines.extend(f"- #{issue}" for issue in completed)
    else:
        lines.append("None")

    lines.extend(
        [
            "",
            "## Models Used",
            "| Issue | Task | Cycle | Agent | Model | Reason |",
            "|---|---|---:|---|---|---|",
        ]
    )
    for issue in completed:
        summary = summaries.get(issue, {})
        report = report_for_issue(state, issue, run_root, summary)
        for row in model_rows(issue, report, summary):
            lines.append(
                f"| #{one_line(issue)} | {row['role']} | {row['cycle']} | "
                f"{row['agent']} | {row['model']} | {row['selection_reason']} |"
            )

    lines.extend(["", "## Verification", "| Issue | State | Evidence |", "|---|---|---|"])
    for issue in completed:
        summary = summaries.get(issue, {})
        report = report_for_issue(state, issue, run_root, summary)
        state_text, evidence = verification_state(report, summary)
        lines.append(f"| #{one_line(issue)} | {state_text} | {evidence} |")

    return "\n".join(lines) + "\n"


def render(args: argparse.Namespace) -> int:
    sys.stdout.write(render_pr_body(args.state_file))
    return 0


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="render run-with-it final PR body")
    subparsers = parser.add_subparsers(dest="command", required=True)
    render_parser = subparsers.add_parser("render")
    render_parser.add_argument("--state-file", required=True)
    render_parser.set_defaults(func=render)
    return parser


def main(argv: list[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    sys.exit(main())
